File: tcp_Chat.java
import java.io.*;
        import java.net.*;
        import java.util.*;
        import java.util.concurrent.*;

public class Main {

    private static final int SERVER_PORT = 12345;
    private static Map<String, ClientHandler> clients = new ConcurrentHashMap<>();

    public static void main(String[] args) {
        if (args.length == 0) {
            startServer();
        } else if (args.length == 2) {
            String name = args[1];
            startClient(args[0], name);
        } else {
            System.out.println("Usage: java ChatProgram [server | client <name>]");
        }
    }

    public static void startServer() {
        try (ServerSocket serverSocket = new ServerSocket(SERVER_PORT)) {
            System.out.println("Server started. Listening on port " + SERVER_PORT);

            while (true) {
                Socket clientSocket = serverSocket.accept();
                ClientHandler handler = new ClientHandler(clientSocket);
                new Thread(handler).start();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static void startClient(String serverAddress, String name) {
        try (Socket socket = new Socket(serverAddress, SERVER_PORT)) {
            System.out.println("Connected to server.");

            BufferedReader reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
            PrintWriter writer = new PrintWriter(socket.getOutputStream(), true);

            writer.println("register " + name);

            new Thread(() -> {
                try {
                    String line;
                    while ((line = reader.readLine()) != null) {
                        System.out.println(line);
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }).start();

            BufferedReader consoleReader = new BufferedReader(new InputStreamReader(System.in));
            String input;
            while ((input = consoleReader.readLine()) != null) {
                String[] parts = input.split(" ", 3);
                if (parts.length == 3 && parts[0].equals("send")) {
                    writer.println("send " + parts[1] + " " + parts[2]);
                } else {
                    System.out.println("Invalid command. Usage: send <recipient> <message>");
                }
            }

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    static class ClientHandler implements Runnable {
        private Socket socket;
        private String name;
        private BufferedReader reader;
        private PrintWriter writer;

        public ClientHandler(Socket socket) throws IOException {
            this.socket = socket;
            this.reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
            this.writer = new PrintWriter(socket.getOutputStream(), true);
        }

        public void run() {
            try {
                String line;
                while ((line = reader.readLine()) != null) {
                    String[] parts = line.split(" ", 3);
                    if (parts[0].equals("register")) {
                        name = parts[1];
                        clients.put(name, this);
                        System.out.println(name + " registered.");
                    } else if (parts[0].equals("send")) {
                        String recipient = parts[1];
                        String message = parts[2];
                        sendMessage(recipient, name + ": " + message);
                    }
                }
            } catch (IOException e) {
                e.printStackTrace();
            } finally {
                if (name != null) {
                    clients.remove(name);
                }
                try {
                    socket.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }

        private void sendMessage(String recipient, String message) {
            ClientHandler clientHandler = clients.get(recipient);
            if (clientHandler != null) {
                clientHandler.writer.println(message);
            } else {
                writer.println("User " + recipient + " not found.");
            }
        }
    }
}

File: tcpchat.py
import socket
import sys
import threading

# Dictionary to store registered clients: {name: (ip, port)}
clients = {}

def handleClient(c_sock, c_address):
    with c_sock:
        while True:
            message = c_sock.recv(1024).decode().rstrip()
            if not message:
                break
            print(f"Received message from {c_address}: {message}")
            if message.startswith("register"):
                parts = message.split()
                if len(parts) == 4:
                    name = parts[1]
                    ip = parts[2]
                    port = int(parts[3])
                    clients[name] = (ip, port)
                    print(f"Registered {name} with IP {ip} and port {port}")
                    print(f"Current clients: {clients}")
            elif message.startswith("send"):
                parts = message.split(' ', 2)
                if len(parts) == 3:
                    target_name = parts[1]
                    msg = parts[2]
                    if target_name in clients:
                        target_ip, target_port = clients[target_name]
                        sendMessage(target_ip, target_port, f"{target_name}: {msg}")
                    else:
                        print(f"Client {target_name} not found")
                else:
                    print("Unknown message format:", message)

def sendMessage(ip, port, message):
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as c_sock:
        try:
            c_sock.connect((ip, port))
            c_sock.sendall(message.encode())
        except ConnectionRefusedError:
            print(f"Failed to send message to {ip}:{port}. Connection refused.")

def startServer(port):
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s_sock:
        s_sock.bind(('0.0.0.0', port))
        s_sock.listen()
        print(f"Server listening on port {port}")
        while True:
            c_sock, c_address = s_sock.accept()
            threading.Thread(target=handleClient, args=(c_sock, c_address), daemon=True).start()

def registerClient(server_ip, server_port, client_name, client_port):
    message = f"register {client_name} {server_ip} {client_port}"
    sendToServer(server_ip, server_port, message)

    # Start listening on the client port after registration
    threading.Thread(target=startClientListener, args=(client_port,), daemon=True).start()

def startClientListener(client_port):
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as c_sock:
        c_sock.bind(('0.0.0.0', client_port))
        c_sock.listen()
        print(f"Client listening on port {client_port}")
        while True:
            conn, addr = c_sock.accept()
            threading.Thread(target=handleClientMessage, args=(conn, addr), daemon=True).start()

def handleClientMessage(conn, addr):
    with conn:
        while True:
            message = conn.recv(1024).decode().rstrip()
            if not message:
                break
            print(f"Received message from {addr}: {message}")

def sendToServer(server_ip, server_port, message):
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as c_sock:
        try:
            c_sock.connect((server_ip, server_port))
            c_sock.sendall(message.encode())
        except ConnectionRefusedError:
            print(f"Failed to connect to server at {server_ip}:{server_port}. Connection refused.")

def main():
    if len(sys.argv) < 3:
        print("Usage:")
        print("For server: python tcpchat.py server <port>")
        print("For client: python tcpchat.py client <server_ip> <server_port> <client_name> [<client_port>]")
        sys.exit()

    if sys.argv[1] == "server":
        port = int(sys.argv[2])
        startServer(port)
    elif sys.argv[1] == "client":
        if len(sys.argv) < 5:
            print("Usage: python tcpchat.py client <server_ip> <server_port> <client_name> [<client_port>]")
            sys.exit()
        
        server_ip = sys.argv[2]
        server_port = int(sys.argv[3])
        client_name = sys.argv[4]
        client_port = int(sys.argv[5]) if len(sys.argv) > 5 else 0
        
        if client_port == 0:
            print("Using random client port.")
        
        registerClient(server_ip, server_port, client_name, client_port)
        
        while True:
            target_name = input("Enter target name (or 'exit' to quit): ")
            if target_name.lower() == 'exit':
                break
            message = input("Enter message: ")
            sendMessage(server_ip, server_port, f"send {target_name} {message}")

if __name__ == '__main__':
    main()
